cmd/kubeadm/app/util/users/users_linux.go

```diff
@@ -0,0 +1,602 @@
+)
+
+// EntryMap holds a map of user or group entries.
+type EntryMap struct {
+	entries map[string]*entry
+}
+
+// entry is a structure that holds information about a UNIX user or group.
+// It partialially conforms parsing of both users from /etc/passwd and groups from /etc/group.
+type entry struct {
+	name      string
+	id        int64
+	gid       int64
+	userNames []string
+	shell     string
+}
+
+// limits is used to hold information about the minimum and maximum system ranges for UID and GID.
+type limits struct {
+	minUID, maxUID, minGID, maxGID int64
+}
+
+const (
+	// these are constants used when parsing /etc/passwd or /etc/group in terms of how many
+	// fields and entry has.
+	totalFieldsGroup = 4
+	totalFieldsUser  = 7
+
+	// klogLevel holds the klog level to use for output.
+	klogLevel = 5
+
+	// noshell holds a path to a binary to disable shell login.
+	noshell = "/bin/false"
+)
+
+var (
+	// these entries hold the users and groups to create as defined in:
+	// https://git.k8s.io/enhancements/keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane
+	usersToCreateSpec = []*entry{
+		{name: "kubeadm-etcd"},
+		{name: "kubeadm-kas"},
+		{name: "kubeadm-kcm"},
+		{name: "kubeadm-ks"},
+	}
+	groupsToCreateSpec = []*entry{
+		{name: "kubeadm-etcd", userNames: []string{"kubeadm-etcd"}},
+		{name: "kubeadm-kas", userNames: []string{"kubeadm-kas"}},
+		{name: "kubeadm-kcm", userNames: []string{"kubeadm-kcm"}},
+		{name: "kubeadm-ks", userNames: []string{"kubeadm-ks"}},
+		{name: "kubeadm-sa-key-readers", userNames: []string{"kubeadm-kas", "kubeadm-kcm"}},
+	}
+
+	// defaultLimits holds the default limits in case values are missing in /etc/login.defs
+	defaultLimits = &limits{minUID: 100, maxUID: 999, minGID: 100, maxGID: 999}
+)
+
+// ID returns the ID for an entry based on the entry name.
+// In case of a user entry it returns the user UID.
+// In case of a group entry it returns the group GID.
+// It returns false as the first returned argument if no such entry exists.
+func (u *EntryMap) ID(name string) (bool, *int64) {
+	entry, ok := u.entries[name]
+	if !ok {
+		return false, nil
+	}
+	id := entry.id
+	return true, &id
+}
+
+// String convers an EntryMap object to readable string.
+func (u *EntryMap) String() string {
+	lines := make([]string, 0, len(u.entries))
+	for k, e := range u.entries {
+		lines = append(lines, fmt.Sprintf("%s{%d,%d};", k, e.id, e.gid))
+	}
+	sort.Strings(lines)
+	return strings.Join(lines, "")
+}
+
+// AddUsersAndGroups adds the managed users and groups to the files specified
+// by pathUsers and pathGroups. It uses the file specified with pathLoginDef to
+// determine limits for UID and GID. If managed users and groups exist in these files
+// validation is performed on them. The function returns a pointer to a Users object
+// that can be used to return UID and GID of managed users.
+func AddUsersAndGroups(pathLoginDef, pathUsers, pathGroups string) (*EntryMap, *EntryMap, error) {
```

Comment:
note, this returns two separate maps of users and groups because Linux has no guarantee that UID and GID for a new user will match in value.
this differs to: https://github.com/kubernetes/kubernetes/pull/102195

Reply:
a unified struct that is returned makes sense.

Reply:
i'd like to keep `entry` private, so i ended up with this:
```
// EntryMap holds a map of user or group entries.
type EntryMap struct {
	entries map[string]*entry
}

// UsersAndGroups is a structure that holds entry maps of users and groups.
// It is returned by AddUsersAndGroups.
type UsersAndGroups struct {
	// Users is an entry map of users.
	Users *EntryMap
	// Groups is an entry map of groups.
	Groups *EntryMap
}
```

this allows for:
```
ug, err := AddUsersAndGroups(...)
...
ug.Users.ID("foo")
```